forte/modules/objects_factory_psi4.py

```diff
@@ -135,55 +130,85 @@ def prepare_forte_objects(self, data, **kwargs):
-
         Parameters
         ----------
         data : ForteData
             The ForteData object containing options and molecule
-        kwargs : dict
-            Keyword arguments
         """
-        ref_wfn = kwargs.get("ref_wfn", None)
-        if ref_wfn:
-            return ref_wfn, False
-        # if no ref_wfn provided, compute it
-        ref_type = data.options.get_str("REF_TYPE")
-        p4print(f"\n  No reference wavefunction provided. Computing {ref_type} orbitals using Psi4...\n")
-        # Decide whether to run SCF or MCSCF based on REF_TYPE
-        job_type = data.options.get_str("JOB_TYPE")
-        do_mcscf = job_type in ["CASSCF", "MCSCF_TWO_STEP"] or data.options.get_bool("MCSCF_REFERENCE")
-        ref_wfn = self.run_psi4(ref_type, data.molecule, not do_mcscf, **kwargs)
-        return ref_wfn, True
+        #  Here we handle twp different scenarios:
+        #  Case | ref_wfn | Action
+        #  -------------------------------------
+        #  1    | None    | Run new SCF or MCSCF from Psi4 and return the wavefunction
+        #  2    | Passed  | Use ref_wfn
+
+        if not ref_wfn:
+            ref_type = data.options.get_str("REF_TYPE")
+            job_type = data.options.get_str("JOB_TYPE")
+            do_mcscf = job_type in ["CASSCF", "MCSCF_TWO_STEP"] or data.options.get_bool("MCSCF_REFERENCE")
+            ref_wfn = self.run_psi4(ref_type, data.molecule, not do_mcscf, **kwargs)
+            return ref_wfn
+        elif ref_wfn:
+            # to test if we can use the orbitals directly, we compare the SO overlap matrices
+            mol_S = psi4.core.Wavefunction.build(data.molecule, kwargs["basis"]).S()
+            ref_S = ref_wfn.S()
+
+            # catch the case where the number of irreps is different
+            if mol_S.nirrep() != ref_S.nirrep():
+                raise ValueError(
+                    "Different number of irreps in the overlap matrices of the reference wavefunction and the molecule."
+                )
+
+            # if the dimensions are the same, compute the difference between the two matrices
+            if mol_S.rowdim() == ref_S.rowdim() and mol_S.coldim() == ref_S.coldim():
+                diff_S = mol_S.clone()
+                diff_S.subtract(ref_S)
+                norm = diff_S.rms()
+                # if the norm of the difference is less than 1e-6, we can use the orbitals directly
+                if norm < 1e-10:
+                    return ref_wfn
+
+            # compute a new Wavefunction object
+            new_wfn = self.run_psi4("scf", data.molecule, False, **kwargs)
+
+            pCa = new_wfn.basis_projection(ref_wfn.Ca(), ref_wfn.nmopi(), ref_wfn.basisset(), new_wfn.basisset())
+            pCb = new_wfn.basis_projection(ref_wfn.Cb(), ref_wfn.nmopi(), ref_wfn.basisset(), new_wfn.basisset())
+            new_wfn.guess_Ca(pCa)
+            new_wfn.guess_Cb(pCb)
```

Comment:
These are passed to `guess_Ca_` and `guess_Cb_` in SCF, but I thought we read orbitals from `Ca_`.

Reply:
Thanks for catching this!